
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://eaton-lab/twig/quick-guide/">
      
      
      
      
        
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.1">
    
    
      
        <title>quick guide - twig</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.484c7ddc.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#quick-guide" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="twig" class="md-header__button md-logo" aria-label="twig" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            twig
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              quick guide
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="twig" class="md-nav__button md-logo" aria-label="twig" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    twig
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Welcome to twig
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    installation
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../macse-prep/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    macse-prep
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../macse-align/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    macse-align
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../macse-refine/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    macse-refine
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../tree-filter/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    tree-filter
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../csubst/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    csubst
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#example-dataset" class="md-nav__link">
    <span class="md-ellipsis">
      
        Example dataset
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#the-fastest-guide" class="md-nav__link">
    <span class="md-ellipsis">
      
        The fastest guide
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sequence-orthology" class="md-nav__link">
    <span class="md-ellipsis">
      
        Sequence orthology
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Sequence orthology">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#fetch-transcriptome-datasets" class="md-nav__link">
    <span class="md-ellipsis">
      
        Fetch transcriptome datasets
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#preprocess-data" class="md-nav__link">
    <span class="md-ellipsis">
      
        Preprocess data
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#homology-scores" class="md-nav__link">
    <span class="md-ellipsis">
      
        Homology scores
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#homology-filtering" class="md-nav__link">
    <span class="md-ellipsis">
      
        Homology filtering
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#orthogroup-identification" class="md-nav__link">
    <span class="md-ellipsis">
      
        Orthogroup identification
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#orthogroup-refinement" class="md-nav__link">
    <span class="md-ellipsis">
      
        Orthogroup refinement
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#orthogroup-sequence-trimming-alignment" class="md-nav__link">
    <span class="md-ellipsis">
      
        Orthogroup sequence trimming &amp; alignment
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tree-analyses" class="md-nav__link">
    <span class="md-ellipsis">
      
        Tree analyses
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#evolutionary-analyses" class="md-nav__link">
    <span class="md-ellipsis">
      
        Evolutionary analyses
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utilities" class="md-nav__link">
    <span class="md-ellipsis">
      
        Utilities
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="quick-guide">quick guide</h1>
<h2 id="example-dataset">Example dataset</h2>
<p>...</p>
<h2 id="the-fastest-guide">The fastest guide</h2>
<p>You can perform an ortholab analysis to assemble and analyze orthogroups
using the all-in-one analysis command <code>orthogroup run</code>. This takes a large
number of options that can be used to specify the run. It will identify
orthology among sequences to group them into orthogroups; align and trim 
orthogroup sequences; infer gene trees; infer a species tree; root gene 
trees and species trees; and report a number of evolutionary analyses of
orthogroups.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="c1"># run the full analysis in one or few commands</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>ortholab<span class="w"> </span>run<span class="w"> </span>data/*.faa<span class="w"> </span>
</code></pre></div>
<!-- ## Orthogroup identification -->
<h2 id="sequence-orthology">Sequence orthology</h2>
<p>A primary goal of ortholab is to identify homology among sequences both
within and between biological samples to enable evolutionary analyses.</p>
<h3 id="fetch-transcriptome-datasets">Fetch transcriptome datasets</h3>
<p>See the data-guide (link) for tips on assembling transcriptome datasets
for use in ortholab, and/or accessing publicly accessible datasets from
online repositories.</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a><span class="c1"># download an example dataset</span>
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>ortholab<span class="w"> </span>seqs<span class="w"> </span>download<span class="w"> </span>--database<span class="w"> </span>phytozome<span class="w"> </span>--accession<span class="w"> </span><span class="m">23456</span><span class="w"> </span>--all
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a># filetree of example dataset (collapsible)
<a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a>...
</code></pre></div></p>
<h3 id="preprocess-data">Preprocess data</h3>
<p>Many datasets will vary in ... One major aspect is in the sequence labels
store in the protein or cds fasta files (also referred to as fasta headers).</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a><span class="c1"># create a map ..</span>
<a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a>ortholab<span class="w"> </span>seqs<span class="w"> </span>create-relabel-map<span class="w"> </span>data/*/*.proteins.faa.gz<span class="w"> </span>&gt;<span class="w"> </span>map_proteins.tsv
<a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a>
<a id="__codelineno-3-4" name="__codelineno-3-4" href="#__codelineno-3-4"></a><span class="c1"># optionally also crate one for the CDS</span>
<a id="__codelineno-3-5" name="__codelineno-3-5" href="#__codelineno-3-5"></a>ortholab<span class="w"> </span>seqs<span class="w"> </span>create-relabel-map<span class="w"> </span>data/*/*.cds.fna.gz<span class="w"> </span>&gt;<span class="w"> </span>map_cds.tsv
</code></pre></div>
<p>If you create both it is important that the label columns match between the two.
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a># show map TSV collapsible
</code></pre></div></p>
<h3 id="homology-scores">Homology scores</h3>
<p>There are many options that can be implemented in this tool to modify which 
program is used to measure homology, and what the minimum cutoff will be. See
the tools tutorial page (link). This will use the map file to relabel sequences
as they are passed into the blast database.
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>ortholab<span class="w"> </span>seqs<span class="w"> </span>homology-score<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-5-2" name="__codelineno-5-2" href="#__codelineno-5-2"></a><span class="w">    </span>data/*/*.proteins.faa.gz<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-5-3" name="__codelineno-5-3" href="#__codelineno-5-3"></a><span class="w">    </span>--map<span class="w"> </span>map.tsv<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-5-4" name="__codelineno-5-4" href="#__codelineno-5-4"></a><span class="w">    </span>--tool<span class="w"> </span>diamond-blast<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-5-5" name="__codelineno-5-5" href="#__codelineno-5-5"></a><span class="w">    </span>--max-evalue<span class="w"> </span><span class="m">0</span>.001<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-5-6" name="__codelineno-5-6" href="#__codelineno-5-6"></a><span class="w">    </span>--outdir<span class="w"> </span>blastdir/<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-5-7" name="__codelineno-5-7" href="#__codelineno-5-7"></a><span class="w">    </span>--use-labels
</code></pre></div></p>
<p>The blast table format can be modified with options to the tool above,
but we restrict it such that the first X columns will always contain the 
same information, and additional requested features are appended as the
final columns.</p>
<p>From looking at this table you can see that we have scored hits between
sequences sometimes with <em>very</em> low sequence similarity scores. This is
sometimes desirable. In the cases where we do not want to allow such 
deep homology, don't worry, we can filter the hits more strictly in the
next step. However, if you modify run parameters of this step, such as
changing the max-evalue or ... , it can have the effect of making the
homology search run faster.
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a># show blast table collapsible
</code></pre></div></p>
<h3 id="homology-filtering">Homology filtering</h3>
<div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a><span class="c1"># filter homology hits to write a graph</span>
<a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a>ortholab<span class="w"> </span>seqs<span class="w"> </span>homology-graph<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-7-3" name="__codelineno-7-3" href="#__codelineno-7-3"></a><span class="w">    </span>blastdir/*.tsv<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-7-4" name="__codelineno-7-4" href="#__codelineno-7-4"></a><span class="w">    </span>--min-identity<span class="w"> </span><span class="m">0</span>.5<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-7-5" name="__codelineno-7-5" href="#__codelineno-7-5"></a><span class="w">    </span>--max-...<span class="w"> </span><span class="m">5</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-7-6" name="__codelineno-7-6" href="#__codelineno-7-6"></a><span class="w">    </span>--min-...<span class="w"> </span><span class="m">0</span>.1<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-7-7" name="__codelineno-7-7" href="#__codelineno-7-7"></a><span class="w">    </span>...<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-7-8" name="__codelineno-7-8" href="#__codelineno-7-8"></a><span class="w">    </span>--length-normalize<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-7-9" name="__codelineno-7-9" href="#__codelineno-7-9"></a><span class="w">    </span>--require-rbbh<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-7-10" name="__codelineno-7-10" href="#__codelineno-7-10"></a><span class="w">    </span>--outdir<span class="w"> </span>blastdir<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-7-11" name="__codelineno-7-11" href="#__codelineno-7-11"></a><span class="w">    </span>--prefix<span class="w"> </span><span class="nb">test</span>
</code></pre></div>
<p>The graph is represented by pairs of sequence labels and a weight value.
This graph file can be visualized in a tool like ... 
Following the approach used by OrthoMCL and Orthofinder, we will split
this graph into smaller orthogroups using a Markov Clustering method 
in <code>mcl</code>. This takes an inflation parameter <code>-i</code> which ...</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a># show blast table collapsible
</code></pre></div>
<h3 id="orthogroup-identification">Orthogroup identification</h3>
<p>Next we write each group of orthologous sequences (orthogroup) to a file.
This tool will take an orthogroup set split from the graph in the
previous step and write a orthogroup</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a><span class="c1"># filter homology hits to write a graph</span>
<a id="__codelineno-9-2" name="__codelineno-9-2" href="#__codelineno-9-2"></a>ortholab<span class="w"> </span>seqs<span class="w"> </span>homology-split<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-9-3" name="__codelineno-9-3" href="#__codelineno-9-3"></a><span class="w">    </span>--graph<span class="w"> </span>blastdir/homology-graph.tsv<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-9-4" name="__codelineno-9-4" href="#__codelineno-9-4"></a><span class="w">    </span>--databases<span class="w"> </span>blastdir/blast*.tsv<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-9-5" name="__codelineno-9-5" href="#__codelineno-9-5"></a><span class="w">    </span>--map<span class="w"> </span>map.tsv<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-9-6" name="__codelineno-9-6" href="#__codelineno-9-6"></a><span class="w">    </span>--format<span class="w"> </span><span class="o">[</span>json,tsv<span class="o">]</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-9-7" name="__codelineno-9-7" href="#__codelineno-9-7"></a><span class="w">    </span>&gt;<span class="w"> </span>blastdir/orthogroups-i-1.5.tsv
</code></pre></div>
<h3 id="orthogroup-refinement">Orthogroup refinement</h3>
<p>In many other software tools you now simply proceed to use your orthogroup
sequences in downstream analyses and assume that homology was accurately assessed.
However, in <em>ortholab</em> we provide a number of additional tools that allow you to
incorporate information you can learn from downstream analyses to feedback into the
orthology assessment in order to refine your orthogroup assignments. This can be
an iterative process. </p>
<p>The two primary sources of additional information include gene tree relationships
and synteny. 
<div class="highlight"><pre><span></span><code><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a>TODO
</code></pre></div></p>
<h3 id="orthogroup-sequence-trimming-alignment">Orthogroup sequence trimming &amp; alignment</h3>
<p>Before running evolutionary analysis of orthogroup sequences we should filter and align 
them. Filters can be used to skip files that do not have taxa of interest. </p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a><span class="c1"># filter homology hits to write a graph</span>
<a id="__codelineno-11-2" name="__codelineno-11-2" href="#__codelineno-11-2"></a>ortholab<span class="w"> </span>seqs<span class="w"> </span>orthogroup-align<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-11-3" name="__codelineno-11-3" href="#__codelineno-11-3"></a><span class="w">    </span>--orthogroups/*.faa<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-11-4" name="__codelineno-11-4" href="#__codelineno-11-4"></a><span class="w">    </span>--map<span class="w"> </span>map.tsv<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-11-5" name="__codelineno-11-5" href="#__codelineno-11-5"></a><span class="w">    </span>--cds<span class="w"> </span>...<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-11-6" name="__codelineno-11-6" href="#__codelineno-11-6"></a><span class="w">    </span>--align-tool<span class="w"> </span>macse2<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-11-7" name="__codelineno-11-7" href="#__codelineno-11-7"></a><span class="w">    </span>--trim-tool<span class="w"> </span>...<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-11-8" name="__codelineno-11-8" href="#__codelineno-11-8"></a><span class="w">    </span><span class="c1"># require taxon A,B,C (use if --map then use mapped names)</span>
<a id="__codelineno-11-9" name="__codelineno-11-9" href="#__codelineno-11-9"></a><span class="w">    </span>--sample-require<span class="w"> </span>A<span class="w"> </span>B<span class="w"> </span><span class="se">\ </span>
<a id="__codelineno-11-10" name="__codelineno-11-10" href="#__codelineno-11-10"></a><span class="w">    </span><span class="c1"># exlude sequence with too many variant sites, indels?</span>
<a id="__codelineno-11-11" name="__codelineno-11-11" href="#__codelineno-11-11"></a><span class="w">    </span>--sequence-filter<span class="w"> </span>...<span class="w"> </span><span class="se">\ </span><span class="w"> </span><span class="c1"># </span>
</code></pre></div>
<h2 id="tree-analyses">Tree analyses</h2>
<p>Now that we have assembled sequences into orthogroups we can proceed to infer the evolutionary relationships among sequences. </p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a>ortholab<span class="w"> </span>trees<span class="w"> </span>infer-gene-trees<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-12-2" name="__codelineno-12-2" href="#__codelineno-12-2"></a><span class="w">    </span>--model<span class="w"> </span>GTR+G+F4<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-12-3" name="__codelineno-12-3" href="#__codelineno-12-3"></a><span class="w">    </span>--bootstraps<span class="w"> </span><span class="m">100</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-12-4" name="__codelineno-12-4" href="#__codelineno-12-4"></a><span class="w">    </span>--max-copy-number<span class="w"> </span><span class="m">10</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-12-5" name="__codelineno-12-5" href="#__codelineno-12-5"></a><span class="w">    </span>--max-size<span class="w"> </span><span class="m">1000</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-12-6" name="__codelineno-12-6" href="#__codelineno-12-6"></a><span class="w">    </span>--cores<span class="w"> </span><span class="m">40</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-12-7" name="__codelineno-12-7" href="#__codelineno-12-7"></a><span class="w">    </span>--threads<span class="w"> </span><span class="m">10</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-12-8" name="__codelineno-12-8" href="#__codelineno-12-8"></a><span class="w">    </span>--outdir<span class="w"> </span>gene-trees/
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a>
</code></pre></div>
<p>```bash
ortholab trees draw gene-trees/OG1.nwk </p>
<h2 id="evolutionary-analyses">Evolutionary analyses</h2>
<p>...</p>
<h2 id="utilities">Utilities</h2>
<p>...</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "..", "features": ["content.code.copy"], "search": "../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.79ae519e.min.js"></script>
      
    
  </body>
</html>